packaging/build-debian-copyright

```diff
@@ -0,0 +1,233 @@
+
+for package_name, package_data in package_lock.get('packages', {}).items():
+    # Skip root package (empty name), dev dependencies, and peer dependencies
+    if not package_name or package_data.get('dev') or package_data.get('peer'):
+        continue
+
+    # Skip packages filtered out by `tools/node-modules runtime-tar`:
+    # - esbuild: arch dependent, must use distribution packages
+    if re.search(r'.*/@?esbuild(/|$)', package_name):
+        continue
+    # unnecessary binary packages
+    if 'watcher-linux' in package_name or 'sass-embedded' in package_name:
+        continue
+
+    # Get license and copyright info
+    license_text, copyrights = get_package_info(package_name, package_data)
+
+    packages_info[package_name] = (license_text, copyrights)
+
+# Extract all individual licenses for validation
+all_licenses_used: set[str] = set()
+for license_text, _copyrights in packages_info.values():
+    # Extract individual licenses from compound expressions (split on 'and'/'or')
+    parts = re.split(r'\s+(?:and|or)\s+', license_text)
+    all_licenses_used.update(part.strip().strip('()') for part in parts)
+
+# Validate all licenses are defined
+license_ids = template_licenses(template)
+missing_licenses = [lic for lic in all_licenses_used
+                    if lic.lower() not in license_ids and lic != 'UNKNOWN']
+
+if missing_licenses:
+    sys.exit(f"ERROR: The following licenses are used but not defined in {TEMPLATE_FILE}:\n"
+             + '\n'.join(f"  - {lic}" for lic in sorted(missing_licenses))
+             + "\n\nPlease add License: paragraphs for these licenses to the template.")
+
+# Group packages by top-level directory (e.g., @patternfly/react-core → @patternfly, react → react)
+toplevel_groups: dict[str, tuple[str, set[str]]] = {}
+
+for package_name in sorted(packages_info.keys()):
+    license_text, copyrights = packages_info[package_name]
+
+    # Extract top-level directory from package name
+    toplevel = package_name.replace('node_modules/', '', 1).split('/')[0]
+
+    # Group by top-level directory, merging copyrights for packages in the same directory
+    if toplevel in toplevel_groups:
+        existing_license, existing_copyrights = toplevel_groups[toplevel]
+        toplevel_groups[toplevel] = (existing_license, existing_copyrights | copyrights)
+    else:
+        toplevel_groups[toplevel] = (license_text, copyrights.copy())
+
+# Generate paragraphs
+paragraphs: list[str] = []
+for toplevel in sorted(toplevel_groups.keys()):
+    license_text, copyrights = toplevel_groups[toplevel]
+
+    # Use fallback if no copyright found
+    if not copyrights:
+        copyrights = {f"Authors of {toplevel}"}
+
+    copyright_text = '\n '.join(sorted(copyrights))
+
+    # For scoped packages, use wildcard; otherwise use package name directly
+    files_pattern = f"node_modules/{toplevel}/*" if toplevel.startswith('@') else f"node_modules/{toplevel}"
```

Comment:
Probably we want the `*` in all cases anyway, no?  Source files are likely to be in subdirectories...

Reply:
I just re-read https://www.debian.org/doc/packaging-manuals/copyright-format/1.0/#files-field and you are right!